### ISSUE TITLE
Test suite needs to be updated for py.test 3.x

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#5277]
At the moment, our test suite fails if using pytest 3.x. Of course, we're not seeing that in the CI because we use the (old) bundled version of pytest 3.x, so this isn't too urgent. But to see the failures, you can do:

```
python setup.py build_ext --inplace
py.test astropy
```

which gives:

```
========================================================== ERRORS ===========================================================
_____________________________________________________ ERROR collecting  _____________________________________________________
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/_pytest/main.py:570: in gethookproxy
    return self._fs2hookproxy[fspath]
E   KeyError: local('/Users/tom/Dropbox/Code/Astropy/astropy/astropy/table/tests')

During handling of the above exception, another exception occurred:
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/_pytest/config.py:301: in _getconftestmodules
    return self._path2confmods[path]
E   KeyError: local('/Users/tom/Dropbox/Code/Astropy/astropy/astropy/table/tests')

During handling of the above exception, another exception occurred:
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/_pytest/config.py:332: in _importconftest
    return self._conftestpath2mod[conftestpath]
E   KeyError: local('/Users/tom/Dropbox/Code/Astropy/astropy/astropy/table/tests/conftest.py')

During handling of the above exception, another exception occurred:
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/_pytest/config.py:338: in _importconftest
    mod = conftestpath.pyimport()
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
astropy/table/__init__.py:28: in <module>
    from .column import Column, MaskedColumn
astropy/table/column.py:15: in <module>
    from ..utils.console import color_print
astropy/utils/console.py:29: in <module>
    from IPython import get_ipython
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/IPython/__init__.py:48: in <module>
    from .core.application import Application
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/IPython/core/application.py:25: in <module>
    from IPython.core.profiledir import ProfileDir, ProfileDirError
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/IPython/core/profiledir.py:29: in <module>
    class ProfileDir(LoggingConfigurable):
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/IPython/core/profiledir.py:52: in ProfileDir
    `profile` option.""",
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/traitlets/traitlets.py:421: in __init__
    DeprecationWarning, stacklevel=stacklevel)
E   DeprecationWarning: metadata {'config': True} was set from the constructor.  Metadata should be set using the .tag() method, e.g., Int().tag(key1='value1', key2='value2')

During handling of the above exception, another exception occurred:
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/py/_path/common.py:332: in visit
    for x in Visitor(fil, rec, ignore, bf, sort).gen(self):
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/py/_path/common.py:378: in gen
    for p in self.gen(subdir):
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/py/_path/common.py:367: in gen
    dirs = self.optsort([p for p in entries
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/py/_path/common.py:368: in <listcomp>
    if p.check(dir=1) and (rec is None or rec(p))])
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/_pytest/main.py:672: in _recurse
    ihook = self.gethookproxy(path)
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/_pytest/main.py:575: in gethookproxy
    my_conftestmodules = pm._getconftestmodules(fspath)
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/_pytest/config.py:315: in _getconftestmodules
    mod = self._importconftest(conftestpath)
/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/_pytest/config.py:340: in _importconftest
    raise ConftestImportFailure(conftestpath, sys.exc_info())
E   _pytest.config.ConftestImportFailure: DeprecationWarning("metadata {'config': True} was set from the constructor.  Metadata should be set using the .tag() method, e.g., Int().tag(key1='value1', key2='value2')",)
E     File "/Users/tom/Dropbox/Code/Astropy/astropy/astropy/table/__init__.py", line 28, in <module>
E       from .column import Column, MaskedColumn
E     File "/Users/tom/Dropbox/Code/Astropy/astropy/astropy/table/column.py", line 15, in <module>
E       from ..utils.console import color_print
E     File "/Users/tom/Dropbox/Code/Astropy/astropy/astropy/utils/console.py", line 29, in <module>
E       from IPython import get_ipython
E     File "/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/IPython/__init__.py", line 48, in <module>
E       from .core.application import Application
E     File "/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/IPython/core/application.py", line 25, in <module>
E       from IPython.core.profiledir import ProfileDir, ProfileDirError
E     File "/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/IPython/core/profiledir.py", line 29, in <module>
E       class ProfileDir(LoggingConfigurable):
E     File "/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/IPython/core/profiledir.py", line 52, in ProfileDir
E       `profile` option.""",
E     File "/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/traitlets/traitlets.py", line 421, in __init__
E       DeprecationWarning, stacklevel=stacklevel)
================================================== pytest-warning summary ===================================================
WC1 None [pytest] section in setup.cfg files is deprecated, use [tool:pytest] instead.
WC1 None pytest_funcarg__cov: declaring fixtures using "pytest_funcarg__" prefix is deprecated and scheduled to be removed in pytest 4.0.  Please remove the prefix and use the @pytest.fixture decorator instead.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================== 2 pytest-warnings, 1 error in 3.93 seconds =========================================
```

**Comment by pllim** [comment]
Wow...

```
8242 passed, 118 skipped, 37 xfailed, 721 pytest-warnings, 223 error in 308.78 seconds
```

Log attached (no `--remote-data`):
[ztestlog.txt](https://github.com/astropy/astropy/files/445397/ztestlog.txt)

**Comment by pllim** [comment]
Note: For another project using Astropy package template, `.. doctest-skip-all` directive no longer works under pytest 3.x as well.

**Comment by eteq** [comment]
Note that there's a suspicion from #5435 that a change in pytest 3.x altered how defaults are interpreted by `getoption` and similar pytest functions.  This does *not* appear to be documented in the pytest 3.x docs (😡 ), but it may address some of this...

**Comment by bsipocz** [comment]
Appveyor breakage of astropy-helpers is also due to pytest 3.x (https://github.com/astropy/astropy-helpers/issues/265 ). While the workaround there is super simple, I'm thinking about to pin the version number of pytest to the one that gets bundled into astropy to be a bit more self consistent.

**Comment by mhvk** [comment]
Once we fix things, we should perhaps use latest pytest for the allowed-to-fail run (with numpy-dev and remote-data).

**Comment by bsipocz** [comment]
OK, the helpers failure is a red herring, but the pytest behaviour is indeed changed. So the failure was within ci-helpers and that got fixed today.
The new puzzling issue is that astropy-helpers should have never picked up the ci-helpers test file.

**Comment by mhvk** [comment]
While trying to investigate, I wanted to just use regular commands for testing so I could focus on specific packages. This, it turns out, [was foreseen](https://github.com/astropy/astropy/blob/master/astropy/tests/helper.py#L43): one can just do
```
export ASTROPY_USE_SYSTEM_PYTEST=1
python3 setup.py test -P ...
```

**Comment by mhvk** [comment]
There seem to be some different failures:
1. tests using the `yield` structure are deprecated; these are easy to fix, and @kelle started some fixes for `wcs` -- I have an additional one for `units`.
2. for doctests, the output sometimes no longer has line breaks that used to be present.
3. `.. doctest-skip` seems to get ignored (both individual and whole-file).

**Comment by mhvk** [comment]
OK, so with #5678 and #5682 merged, the first item on the above little list is checked off. Now need to understand why the output changes (different line length assumed?) and why `doctest-skip` no longer works...

**Comment by pllim** [comment]
@mhvk , I spent a good few hours trying to understand the doctest failure but got nowhere. I tried inserting print statements but they don't work with pytest, as it suppresses print outputs, I think. There is no traceback, just some obscure deprecation warnings. And no clear documented translation guide for those deprecations. :cold_sweat:

**Comment by astrofrog** [comment]
@pllim - you can run pytest with the ``-s`` option to see the output of print statements.

**Comment by pllim** [comment]
Some warnings are due to a bug in pytest 3.0.5, see pytest-dev/pytest#2118, although that still does not explain why doctest is not skipping some files/snippets as it should. FYI.

**Comment by mhvk** [comment]
I've been trying to narrow down a bit. A first conclusion is that the problem only occurs for text files, not for source files (modules; e.g., `__doctest_skip__` in *.py files does work). A second is that the options we pass on, in particular `NORMALIZE_WHITESPACE` in `pytest_plugins.pytest_configure` does not seem to get used for `*.rst` files.
...
...
...
Looking in more detail, the main problem seems to be that there is no `collect` method defined for [DocTestTextfilePlus](https://github.com/astropy/astropy/blob/master/astropy/tests/pytest_plugins.py#L197). Indeed, by adding a collect method following the pytest `DoctestTextfile` class, I can get the tests to pass! (well, minus some annoying `vo` tests, but what's new...) See #5688.

**Comment by mhvk** [comment]
Closing this, since with #5688 astropy now works well with pytest 3.x.

**Comment by cdeil** [comment]
This just now appeared, I think because our CI switched to the recently released pytest 3.2:
https://travis-ci.org/hipspy/hips/jobs/260162492#L1211
```
INTERNALERROR> AttributeError: 'DocTestTextfilePlus' object has no attribute '_location'
```
Should I file a new issue about this? In which tracker?

**Comment by bsipocz** [comment]
We're on it. Long story short it's not yet clear whether it's a bug in our plugin or in pytest so we need to limit the version number of pytest for now, I'm waiting for the builds to pass on ci-helpers. 

https://github.com/pytest-dev/pytest/pull/2651
